projects/sandbox/analyze/analyze.py

```diff
@@ -455,110 +422,98 @@ def analyze_injections(
                 if normalizer is not None:
                     normalizer.fit(back_y)
 
-                # submit integration job
-                future = process_ex.submit(
-                    integrate,
+                # run integration job in main process
+                t, y, integrated = integrate(
                     injection_y,
                     t,
                     kernel_length=kernel_length,
                     window_length=window_length,
                     normalizer=normalizer,
                 )
-                future.add_done_callback(
-                    lambda f: pbar.update(analyze_task_id, advance=1)
-                )
-
-                integrate_futures[shift] = [future]
-
-                pbar.update(load_task_id, advance=2)
-
-            # as the integration jobs come back
-            # submit jobs to write integration outputs
-
-            # Also, submit jobs to analyze event times
-            # using the integrated outputs
-            write_futures = []
-            characterize_futures = []
-            for shift, (t, y, integrated) in as_completed(integrate_futures):
-
-                # get params file for injections for this timeshift
-                param_file = data_dir / shift / "injection" / "params.h5"
-                with h5py.File(param_file) as f:
 
-                    # TODO: we seem to be systematically off by 1.5 seconds;
-                    # need to find discrepancy
-                    event_times = f["geocent_time"][()] + 1.5
-
-                # restrict event times of interest
-                # to those within this segment, taking
-                # into account normalization
-                segment_event_times = [
-                    time
-                    for time in event_times
-                    if time < event_window[1] and time > event_window[0]
-                ]
+                pbar.update(analyze_task_id, advance=1)
 
+                # submit write job for the integrated output
                 shift_dir = get_write_dir(
                     write_dir, shift, "injection-integrated", norm
                 )
 
                 write_future = thread_ex.submit(
                     write_timeseries,
                     shift_dir,
+                    prefix="integrated",
                     t=t,
                     y=y,
                     integrated=integrated,
                 )
 
-                write_futures.append(future)
                 write_future.add_done_callback(
                     lambda f: pbar.update(write_task_id, advance=1)
                 )
 
+                # get params file for injections for this timeshift
+                param_file = data_dir / shift / "injection" / "params.h5"
+                with h5py.File(param_file) as f:
+                    # TODO: we seem to be systematically off by 1.5 seconds;
+                    # need to find discrepancy
+                    event_times = f["geocent_time"][()] + 1.5
```

Comment:
I think the event times are right, it's the `t` array that gets written from inference that's off. `integrate` adds `kernel_length` to the `t` array because in Tri's original implementation, `t` corresponded to the time at left edge of the kernel, not the right. The current inference code accounts for this, so that should be taken out of `integrate`.

The 0.5 is more interesting: the leading edge of the kernel that gets streamed to the server will always be `fduration / 2` ahead of the leading edge of the kernel that gets passed to the NN due to the center slice in the whitening operation. In principle this means we have a baked-in `fduration / 2` second latency in production, but since the low-latency frames are written in 1s increments this shouldn't matter too much functionally. But here, we should expect the timestamp at which the network starts firing to be `fduration / 2` seconds _after_ the event time. So it might make sense to expose `fduration` as an arg here and subtract that from `t` (or we can leave `t` as-is and include the 0.5s extra latency in the characterized `latency` array)

Reply:
Ah, that all adds up very nicely

Reply:
I'm wondering if it would make more sense to take this into account in the inference code, so that we don't have to remember to perform this shift every time we do some analysis on network outputs.

Reply:
Also, in the current inference code, is the shift by the kernel length taken care of on the `hermes` side?

Reply:
So for the kernel length shift, the inference code actually doesn't do anything. It loads in the `t` array and `x` arrays, downsamples `t` to the inference sampling rate (just a strided slice), and then pairs this up with the array of network responses `y` for each `x`. The key is that each `y` is the network's prediction on a kernel for which the corresponding `x` stream (and timestamp from `t`) are on the right edge of the kernel, and the "shift" really comes from different ways of defining the timestamp of a kernel.  `integrate` adds `kernel_length` _again_ because Tri's code defined `t` differently, so really the fix here is to make `integrate` simpler and more general by taking that extra addition out.

As for the `fduration / 2` shift, that poses a more interesting dilemma, because from the viewpoint of identifying detection latency, that shift is very real: our network will actually fire `fduration / 2`s after the coalescence time, and so you could argue it's worth documenting that latency. However, for reconstructing the event time (or the time of some glitch that causes an erroneous trigger), you're right that it could be obnoxious and laborious to have to account for this every time. If the latency analysis is less important because we'll just take the max of the integrated NN output over some window, then agreed it will make more sense to just account for this in the inference code.

Reply:
Okay yeah totally agree that keeping track of that incurred latency is super important especially once we start running online applications where 0.5 seconds makes a world of difference. For these 'offline' applications, I still think it makes more sense to save to disk the network outputs associated with the times that BBHnet actually uses to make decisions, and treat that extra `fduration / 2` as some additional processing latency.

Reply:
Yeah I think I'm on board with that, sounds good